sql/src/main/java/io/crate/operation/projectors/ShardingUpsertExecutor.java

```diff
@@ -233,13 +233,13 @@ private ShardLocation getShardLocation(String indexName, String id, @Nullable St
         }
     }
 
-    private void addToRequest(TItem item, ShardLocation shardLocation, Map<ShardLocation, TReq> requests) {
+    private synchronized void addToRequest(TItem item, ShardLocation shardLocation, Map<ShardLocation, TReq> requests) {
```

Comment:
@andreidan @mfussenegger pointed out that the whole method should be synchronized, because also the concurrent read/write access on the `requests` map could lead to race conditions.
The performance overhead should be very small because concurrent access happens only in the callback of the create index future.

Reply:
Please use `sychronized(lock) {...}` instead of locking on a method level. It might be marginal here but, generally, we want to isolate only the parts where concurrency is an issue, i.e. the map and the counter.

You still have a performance overhead for synchronizing even if the lock is only ever acquired by one thread at a time.`ConcurrentHashMap` is usually an option to avoid locking but it doesn't work here.

`location` should be volatile, otherwise updates to the variable might not be visible to all threads immediately.

Reply:
> location should be volatile, otherwise updates to the variable might not be visible to all threads immediately.

If it's only accessed within synchronized blocks that's not necessary.

> It might be marginal here  

Seems like there is no difference at all in this case?

Reply:
True, `volatile` is not necessary if the variable is encapsulated in a synchronized block.

Reply:
@mfussenegger @mxm so you want any changes here or not?

Reply:
I'd still be in favor of the isolated locking. `volatile` flag is not necessary as pointed out.

Reply:
@mxm IntelliJ warns that synchronization on a local variable or parameter cannot be guaranteed.

Reply:
@mxm I'm not sure you can reduce the scope in this case, can you? (as all this method does is deal with the requests map and the location counter)
Or are you thinking of not promoting the synchronized method model?

Reply:
@chaudum Synchronizing on a local variable indeed won't work. You will have to use a lock object.

@andreidan You're right, the scope is kind of limited here but it might be extended in the future and synchronizing only the relevant part is a defensive way to avoid too much locking.

There are two places where `addToRequest` is called:

In line 146 in `createRowConsumer` you need:
```java
                synchronized (this) {
                    addToRequest(item, shardLocation, requestsByShard);
                }
``` 

In `getFromPendingToRequestMap` you actually don't need to lock the `addToRequest` because it operates on a local map. However, you will have to use AtomicInteger for the locations.

Just a suggestion, you could also simply lock the whole method.

By the way, is `int` safe here? That means that we can't insert more values than Integer.MAX_VALUE.

Reply:
@mxm I updated the PR with your suggestion to wrap the whole `addToRequest()` method into a `synchronized(this)` block, but this results in a deadlock :-/

Reply:
however, when synchronizing both `addToRequest()` method calls it works again. then I prefer to annotate the method with synchronized. also the AtomicInteger would not be necessary then.

Reply:
hm, travis passes but when I run tests locally they hang.